File: Algorithms CPP/nextPermutation.java
import java.util.*;

class nextPermutation {
  public static void swap(int[] arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }

  public static void reverse(int[] arr, int i, int j) {
    while (i < j) {
      swap(arr, i, j);
      i++;
      j--;
    }
  }

  public static void nextPermutation(int[] nums) {
    if (nums == null || nums.length <= 1)
      return;
    int i = nums.length - 2;
    while (i >= 0 && nums[i] >= nums[i + 1])
      i--;
    if (i >= 0) {
      int j = nums.length - 1;
      while (nums[j] <= nums[i])
        j--;
      swap(nums, i, j);
    }
    reverse(nums, i + 1, nums.length - 1);
  }

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int n = sc.nextInt();
    int[] arr = new int[n];
    for (int i = 0; i < n; i++) {
      arr[i] = sc.nextInt();
    }
    nextPermutation(arr);
    for (int i = 0; i < n; i++) {
      System.out.print(arr[i]);
    }
  }
}

File: Leetcode Problems/NextPermutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n=nums.size();
        if(n==1){
            return;
        }
        int i=1;
        int lastInc=-1;
        while(i<n){
            if(nums[i]>nums[i-1])
                lastInc=i;
            i+=1;
        }
        if(lastInc==-1){
            for(i=0;i<n/2;++i){
                swap(nums[i],nums[n-i-1]);
            }
            return;
        }
        int mn=nums[lastInc];
        int index=lastInc;
        for(i=lastInc;i<n;++i){
            if(nums[i]>nums[lastInc-1] and nums[i]<nums[index])
                index=i;
        }
        swap(nums[lastInc-1],nums[index]);
        sort(nums.begin()+lastInc,nums.end());
        
    }
};
